File: src/main/scala/extracells/integration/mekanism/gas/GasStorageChannel.scala
package extracells.integration.mekanism.gas

import appeng.api.config.{Actionable, PowerMultiplier}
import appeng.api.networking.energy.IEnergySource
import appeng.api.networking.security.IActionSource
import appeng.api.storage.IMEInventory
import appeng.api.storage.data.IItemList
import extracells.api.gas.{IAEGasStack, IGasStorageChannel}
import extracells.util.GasUtil
import io.netty.buffer.ByteBuf
import mekanism.api.gas.{Gas, GasRegistry, GasStack}
import net.minecraftforge.fluids.{Fluid, FluidStack}


object GasStorageChannel extends IGasStorageChannel {
  override def poweredExtraction(energy: IEnergySource, cell: IMEInventory[IAEGasStack], t: IAEGasStack, source: IActionSource): IAEGasStack = {
    var cellContains = cell.extractItems(t.copy(), Actionable.SIMULATE, source)
    val count = if (cellContains != null) cellContains.getStackSize else 0
    val maxEnergy = energy.extractAEPower(count, Actionable.SIMULATE, PowerMultiplier.CONFIG)
    val toExtract = Math.min(count, maxEnergy.toLong)
    if (toExtract > 0) {
      energy.extractAEPower(count, Actionable.MODULATE, PowerMultiplier.CONFIG)
      cellContains = cellContains.copy().setStackSize(toExtract)
      cell.extractItems(cellContains, Actionable.MODULATE, source)
    } else
      null
  }


  override def readFromPacket(byteBuf: ByteBuf): IAEGasStack = new AEGasStack(byteBuf)

  override def createList(): IItemList[IAEGasStack] = new GasList

  override def poweredInsert(energy: IEnergySource, cell: IMEInventory[IAEGasStack], t: IAEGasStack, source: IActionSource): IAEGasStack = {
    var cellFree = cell.injectItems(t.copy(), Actionable.SIMULATE, source)
    val count = if (cellFree != null) t.getStackSize - cellFree.getStackSize else t.getStackSize
    val maxEnergy = energy.extractAEPower(count, Actionable.SIMULATE, PowerMultiplier.CONFIG)
    val toInsert = Math.min(count, maxEnergy.toLong)
    if (toInsert > 0) {
      energy.extractAEPower(count, Actionable.MODULATE, PowerMultiplier.CONFIG)
      cellFree = cellFree.copy().setStackSize(toInsert)
      cell.injectItems(cellFree, Actionable.MODULATE, source)
    } else
      null
  }

  override def createStack(o: scala.Any): IAEGasStack = {
    o match {
      case gas :Gas => new AEGasStack(new GasStack(gas, 1000))
      case gasStack :GasStack => new AEGasStack(gasStack)
      case gasStack :AEGasStack => new AEGasStack(gasStack)
      case fluid :Fluid => {
        if(GasUtil.isGas(fluid))
          new AEGasStack(GasUtil.getGasStack(new FluidStack(fluid, 1000)))
        else{
          val gas = GasRegistry.getGas(fluid)
          if (gas != null) new AEGasStack(new GasStack(gas, 1000))
          else null
        }
      }
      case fluidStack :FluidStack => {
        if (fluidStack.getFluid == null) null
        else if(GasUtil.isGas(fluidStack)) new AEGasStack(GasUtil.getGasStack(fluidStack))
        else{
          val gas = GasRegistry.getGas(fluidStack.getFluid)
          if (gas != null) new AEGasStack(new GasStack(gas, fluidStack.amount))
          else null
        }
      }
      case _ => null
    }
  }
}

File: src/main/scala/extracells/integration/mekanism/gas/GasStorageChannel.java
package extracells.integration.mekanism.gas;


import appeng.api.networking.energy.IEnergySource;
import appeng.api.networking.security.IActionSource;
import appeng.api.storage.IMEInventory;
import appeng.api.storage.data.IItemList;
import appeng.util.Platform;
import com.google.common.base.Preconditions;
import extracells.api.gas.IAEGasStack;
import extracells.api.gas.IGasStorageChannel;
import extracells.util.GasUtil;
import io.netty.buffer.ByteBuf;
import mekanism.api.gas.Gas;
import mekanism.api.gas.GasRegistry;
import mekanism.api.gas.GasStack;
import net.minecraftforge.fluids.Fluid;
import net.minecraftforge.fluids.FluidStack;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.io.IOException;

import net.minecraft.nbt.NBTTagCompound;


public class GasStorageChannel implements IGasStorageChannel {

    @Override
    public int getUnitsPerByte()
    {
        return 8000;
    }

    @Nonnull
    @Override
    public IItemList<IAEGasStack> createList() {
        return new GasList();
    }

    @Nullable
    @Override
    public IAEGasStack readFromPacket(@Nonnull ByteBuf byteBuf) throws IOException {
        return new AEGasStack(byteBuf);
    }

    @Nullable
    @Override
    public IAEGasStack createFromNBT(@Nonnull NBTTagCompound nbtTagCompound) {
        return new AEGasStack(nbtTagCompound);
    }

    @Nullable
    @Override
    public IAEGasStack createStack(@Nonnull Object o) {
        if (o instanceof Gas) {
            return new AEGasStack(new GasStack((Gas) o, 1000));
        } else if (o instanceof GasStack) {
            return new AEGasStack((GasStack) o);
        } else if (o instanceof AEGasStack) {
            return new AEGasStack((AEGasStack) o);
        } else if (o instanceof Fluid) {
            Fluid fluid = (Fluid) o;
            if (GasUtil.isGas(fluid)) {
                return new AEGasStack(GasUtil.getGasStack(new FluidStack(fluid, 1000)));
            } else {
                Gas gas = GasRegistry.getGas(fluid);
                if (gas != null) {
                    return new AEGasStack(new GasStack(gas, 1000));
                }
            }
        } else if (o instanceof FluidStack) {
            FluidStack fluidStack = (FluidStack) o;
            if (fluidStack.getFluid() == null)
                return null;
            if (GasUtil.isGas(fluidStack))
                return new AEGasStack(GasUtil.getGasStack(fluidStack));
            return this.createStack(fluidStack.getFluid());
        }
        return null;
    }
}
